traits/trait_list_object.py

```diff
@@ -53,460 +88,750 @@ def __init__(self, index=0, removed=None, added=None):
+        norm_index = self._normalize_index(index)
+
+        removed = super().pop(index)
+
+        self.notify(norm_index, removed, added)
+
+        return removed
 
     def remove(self, value):
-        trait = getattr(self, "trait", None)
-        if trait is None:
-            list.remove(self, value)
-            return
-        if trait.minlen < len(self):
-            try:
-                index = self.index(value)
-                removed = [self[index]]
-            except Exception:
-                pass
-
-            list.remove(self, value)
-
-            if self.name_items is not None:
-                self._send_trait_items_event(
-                    self.name_items, TraitListEvent(index, removed)
-                )
-        elif len(self) == 0:
-            # Let whatever system error (ValueError) should be raised be
-            # raised.
-            list.remove(self, value)
-        else:
-            self.len_error(len(self) - 1)
+        """ Remove first occurrence of value.
+
+        Parameters
+        ----------
+        value : Any
+            A value contained in the list.
+
+        Raises
+        ------
+        ValueError
+            If the value is not present.
+
+        Notes
+        -----
+        Notification:
+            index : int
+                The index between 0 and len(self).
+            added : Undefined
+                Will be Undefined
+            removed : item
+                The removed item.
+
+        """
+        index = self.index(value)
+        added = self.validate(index, value, Undefined)
+
+        super().remove(value)
+
+        self.notify(index, value, added)
 
     def sort(self, key=None, reverse=False):
-        removed = self[:]
-        list.sort(self, key=key, reverse=reverse)
-        if (
-            getattr(self, "name_items", None) is not None
-            and getattr(self, "trait", None) is not None
-        ):
-            self._send_trait_items_event(
-                self.name_items, TraitListEvent(0, removed, self[:])
-            )
+        """ Sort the items in the list in ascending order, *IN PLACE*.
+        A custom key function can be supplied to customize
+        the sort order, and the reverse flag can be set to request the result
+        in descending order.
+
+        Parameters
+        ----------
+        key : Callable
+            Custom function.
+        reverse : bool
+            If true, the resulting list will be sorted in descending order.
+
+        Notes
+        -----
+        Notification:
+            index : slice
+                The slice between 0 and len(self)
+            added : list
+                Will be [].
+            removed : Undefined
+                Will be Undefined
+
+        """
+        self[:] = sorted(self, key=key, reverse=reverse)
+        index = slice(0, len(self), None)
+
+        # Notes is not fired if added == removed, so make them unequal
+        added = []
+        removed = Undefined
```

Comment:
Removed `Undefined`

> The original `TraitListObject` behaves differently: `removed` is the original content, and `added` is the new content. Would this be a backward incompatible change?

Hmm, It would, didn't notice that, made the change. Thanks!